These scripts were written at UC Berkeley MVZ and/or CGRL and used for analyzing my hummingbird exon sequence capture data. These are the versions of the scripts I used in my analyses. 

If you use these scripts, please cite appropriately.

- Sonal Singhal and Ke Bi wrote: `1-ScrubReads_d`, `2-GenerateAssembliesPhylo`, and `3-FindingTargetsV9`
  - Documentation for these steps:
    - [clean reads](https://github.com/CGRL-QB3-UCBerkeley/seqCapture/wiki/Cleaning-raw-reads)
    - [assemble reads](https://github.com/CGRL-QB3-UCBerkeley/seqCapture/wiki/De-novo-assembly-population-genetics)
    - [find targets](https://github.com/CGRL-QB3-UCBerkeley/seqCapture/wiki/Exon-capture-designed-based-on-de-novo-assembled-transcriptomes,-popgen-project)
- Ke Bi additionally wrote: `5-Alignment`, `6-PopStats`, `6-exonCaptureEvaluation`, `average_theta.pl`
- `SNPcleaner225.pl`: [reference](https://onlinelibrary.wiley.com/doi/full/10.1111/mec.12516?casa_token=BaQ9-ZRwsoQAAAAA%3A_nS367XIrr5McCPKGGeYege0GfGmIJnwhNiDAot-KvtWf0F01eH6mpwG6AIJMpe0tWzOV6Vo31XN4YA)

More up-to-date and other scripts are available on the [CGRL-QB3-UCBerkeley Github](https://github.com/CGRL-QB3-UCBerkeley?tab=repositories&q=&type=&language=). 

